From 678fd77f4c2d57b2a5d602acfe7b54722d14d30b Mon Sep 17 00:00:00 2001
From: Ray Zhang <sgzhang@google.com>
Date: Mon, 10 Oct 2022 04:55:49 +0000
Subject: [PATCH 854/905] [Backport] mtd: mtdoops: panic caused mtdoops to call
 mtdoops_erase function immediately

mainline inclusion
from mainline-v6.2-rc1
commit 7cc84e0e07d063c246809ea78f51607f858766bf
category: bugfix
bugzilla: 188253

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7cc84e0e07d063c246809ea78f51607f858766bf

--------------------------------

The panic function disables the local interrupts, preemption, and all
other processors. When the invoked mtdoops needs to erase a used page,
calling schedule_work() to do it will not work. Instead, just call
mtdoops_erase function immediately.

Tested:
~# echo c > /proc/sysrq-trigger
[  171.654759] sysrq: Trigger a crash
[  171.658325] Kernel panic - not syncing: sysrq triggered crash
......
[  172.406423] mtdoops: not ready 34, 35 (erase immediately)
[  172.432285] mtdoops: ready 34, 35
[  172.435633] Rebooting in 10 seconds..

Signed-off-by: Ray Zhang <sgzhang@google.com>
Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>
Link: https://lore.kernel.org/linux-mtd/20221010045549.2221965-4-sgzhang@google.com

conflicts:
	drivers/mtd/mtdoops.c

Signed-off-by: Yi Yang <yiyang13@huawei.com>
---
 drivers/mtd/mtdoops.c | 24 ++++++++++++++++--------
 1 file changed, 16 insertions(+), 8 deletions(-)

diff --git a/drivers/mtd/mtdoops.c b/drivers/mtd/mtdoops.c
index c6842a73bc83..c58605ea6140 100644
--- a/drivers/mtd/mtdoops.c
+++ b/drivers/mtd/mtdoops.c
@@ -160,7 +160,7 @@ static void mtdoops_workfunc_erase(struct work_struct *work)
 	mtdoops_erase(cxt);
 }
 
-static void mtdoops_inc_counter(struct mtdoops_context *cxt)
+static void mtdoops_inc_counter(struct mtdoops_context *cxt, int panic)
 {
 	cxt->nextpage++;
 	if (cxt->nextpage >= cxt->oops_pages)
@@ -170,12 +170,20 @@ static void mtdoops_inc_counter(struct mtdoops_context *cxt)
 		cxt->nextcount = 0;
 
 	if (page_is_used(cxt, cxt->nextpage)) {
-		schedule_work(&cxt->work_erase);
-		return;
+		printk(KERN_DEBUG "mtdoops: not ready %d, %d (erase %s)\n",
+				cxt->nextpage, cxt->nextcount,
+				panic ? "immediately" : "scheduled");
+		if (panic) {
+			/* In case of panic, erase immediately */
+			mtdoops_erase(cxt);
+		} else {
+			/* Otherwise, schedule work to erase it "nicely" */
+			schedule_work(&cxt->work_erase);
+		}
+	} else {
+		printk(KERN_DEBUG "mtdoops: ready %d, %d (no erase)\n",
+				cxt->nextpage, cxt->nextcount);
 	}
-
-	printk(KERN_DEBUG "mtdoops: ready %d, %d (no erase)\n",
-			cxt->nextpage, cxt->nextcount);
 }
 
 static void mtdoops_write(struct mtdoops_context *cxt, int panic)
@@ -207,7 +215,7 @@ static void mtdoops_write(struct mtdoops_context *cxt, int panic)
 	mark_page_used(cxt, cxt->nextpage);
 	memset(cxt->oops_buf, 0xff, record_size);
 
-	mtdoops_inc_counter(cxt);
+	mtdoops_inc_counter(cxt, panic);
 }
 
 static void mtdoops_workfunc_write(struct work_struct *work)
@@ -268,7 +276,7 @@ static void find_next_position(struct mtdoops_context *cxt)
 		cxt->nextcount = maxcount;
 	}
 
-	mtdoops_inc_counter(cxt);
+	mtdoops_inc_counter(cxt, 0);
 }
 
 static void mtdoops_do_dump(struct kmsg_dumper *dumper,
-- 
2.17.1

